Lazy load images

nicl opened the pull request:
## What does this change?

Lazy load images, excluding images in the top two containers on fronts. Lighthouse suggests this is the largest potential win (3s off TTI on a mid Android device with a fast 3g connection for our UK front). BBC and NYTimes (among others) use this approach on fronts to improve performance.

A library called lozad.js is used here (although we could inline it). It is < 1kb and runs during standard (as loading images is considered essential!). Images wrapped in noscript are used to support browsers with Javascript disabled (although they will not see the performance benefits).

It turns out we already include the necessary polyfills (IntersectionObserver) for this to work on older browsers. These browsers will see slightly worse performance though, as the polyfill behavious is slower, though I haven't quantified this.

## What is the value of this and can you measure success?

The advantages of lazy-loading here are:

* reduced initial page load
* improved time to interactive
* cost savings on bandwidth for us (as users may not scroll the entire page)

The cons are:

* additional complexity in code
* potential for flashing of images when scrolling faster than images can load (either because connection is slow, or user is scrolling very fast - rootMargin property can address this somewhat)
* worse experience for users with JS disabled (they currently don't get the full range of sources we normally support in picture but only a fallback)
* potential to break image loading if JS thread is blocked on long-running code

The more images on the page, and the larger they are, the bigger the performance wins. So larger fronts are where the main wins here are.

Webpagetest data below.

**Without lazy-loading**

![screenshot 2018-12-05 at 15 41 42](https://user-images.githubusercontent.com/858402/49525847-7f5f1f00-f8a6-11e8-8610-4b6a2f820e66.png)

**With lazy-loading**

![screenshot 2018-12-05 at 15 53 33](https://user-images.githubusercontent.com/858402/49525795-6c4c4f00-f8a6-11e8-8410-76254b84e209.png)

Note, the document complete time is significantly improved, and the page weight is considerably down. I'm not sure their TTI measure is wholly reliable as across runs it fluctuated considerably so probably just ignore that for now. The change in request count is evidence of the lazy-loading, but this was also confirmed by looking explicitly at image loading.

## Questions/thoughts

* should we support articles as well as fronts? If so, which artlces?

There are potentially lots of edge cases here around different article types and product requirements, so it may be simpler to just support fronts - arguably the biggest gains are here anyway. As things stand, these are not really considered but just default to lazy load.

*UPDATE: I've removed lazy loading on articles for now as I think it needs further discussion.*

* should the default for images be lazy-load?

I set the default to on and the whitelisted exceptions (like initial containers and main images). This helps ensure maximum performance gains, but also involves more risk.

PRBuilds commented:
PRbuilds results:


**Screenshots**
[wide.png](https://s3-eu-west-1.amazonaws.com/prbuilds/PR-20809/screenshots/books%252F2014%252Fmay%252F21%252Fguardian-journalists-jonathan-freedland-ghaith-abdul-ahad-win-orwell-prize-journalism/wide.png) • [desktop.png](https://s3-eu-west-1.amazonaws.com/prbuilds/PR-20809/screenshots/books%252F2014%252Fmay%252F21%252Fguardian-journalists-jonathan-freedland-ghaith-abdul-ahad-win-orwell-prize-journalism/desktop.png) • [tablet.png](https://s3-eu-west-1.amazonaws.com/prbuilds/PR-20809/screenshots/books%252F2014%252Fmay%252F21%252Fguardian-journalists-jonathan-freedland-ghaith-abdul-ahad-win-orwell-prize-journalism/tablet.png) • [mobile.png](https://s3-eu-west-1.amazonaws.com/prbuilds/PR-20809/screenshots/books%252F2014%252Fmay%252F21%252Fguardian-journalists-jonathan-freedland-ghaith-abdul-ahad-win-orwell-prize-journalism/mobile.png)



&#x1F49A; **Exceptions**
[thrown-exceptions.js](https://s3-eu-west-1.amazonaws.com/prbuilds/PR-20809/thrown-exceptions.js)





&#x1F49A; **A11y validation**
[a11y-report.txt](https://s3-eu-west-1.amazonaws.com/prbuilds/PR-20809/a11y-report.txt)



&#x1F49A; **Microdata Validation**
[microdata.txt](https://s3-eu-west-1.amazonaws.com/prbuilds/PR-20809/microdata.txt)

    

**Apache Benchmark Load Testing**
[loadtesting.txt](https://s3-eu-west-1.amazonaws.com/prbuilds/PR-20809/loadtesting.txt)



**LightHouse Reporting**
[lighthouse.html](https://s3-eu-west-1.amazonaws.com/prbuilds/PR-20809/lighthouse.html)


    
--automated message

nicl commented on common/app/views/fragments/image.scala.html:
```diff
@@ -33,33 +35,57 @@
-    @* IE 9 needs this workaround as per https://scottjehl.github.io/picturefill/ *@
-    <!--[if IE 9]><video style="display: none;"><![endif]-->
+@* Nb. assumes lazy loading JS is active *@
+@if(shouldLazyLoad && Switches.LazyLoadImagesSwitch.isSwitchedOn) {
```
Note, the if/else branches here involve a lot of repetition. I'm planning to try and share stuff here a bit but thoughts on this welcome.

nicl commented on common/app/views/fragments/items/elements/facia_cards/image.scala.html:
```diff
@@ -2,27 +2,46 @@
-    <!--[if IE 9]></video><![endif]-->
-    @maybeImageMedia.map(ImgSrc.getFallbackUrl(_)).orElse(maybeSrc).orElse(maybePath).map { src =>
-        <img class="@RenderClasses(classes: _*)" alt="" src="@src"/>
+@if(shouldLazyLoad && Switches.LazyLoadImagesSwitch.isSwitchedOn) {
```
Ditto point above about if/else duplicating too much code here.

nicl commented on common/test/views/support/TemplatesTest.scala:
```diff
@@ -43,7 +43,7 @@ class TemplatesTest extends FlatSpec with Matchers with GuiceOneAppPerSuite {
   }
 
-  "PictureCleaner" should "correctly format inline pictures" in {
+  "PictureCleaner" should "correctly format inline pictures" ignore {
```
TODO fix this.

NataliaLKB commented:
> should we support articles as well as fronts? If so, which artilces?

I think there is the most benefit to Fronts so let's just try it there for now. We can always chose to implement for articles in a later PR

nicl commented:
@NataliaLKB yep, had come to same conclusion. Articles (including galleries, immersives, etc.) requires more thought.

NataliaLKB commented on common/app/views/fragments/items/elements/facia_cards/image.scala.html:
```diff
@@ -2,27 +2,46 @@
+        }
+
+        </picture>
+        <noscript><img class="@RenderClasses(classes: _*)" src="@src" alt=""/></noscript>
```
Why not put a picture element here instead of just an img fallback?

nicl replied:
That is an excellent point!

SiAdcock commented:
> I'm not sure their TTI measure is wholly reliable as across runs it fluctuated considerably so probably just ignore that for now.

I've come to the same conclusion, which is a shame. Perhaps the metrics would be more stable if we disabled for ads for these runs?

GHaberis commented:
@nicl I'd be interested to see how the size of the Document compares with this change? I'm assuming it'll increase now each image also contains a `<noscript><picture></picture></noscript>` element? It'd be good to see by how much.

GHaberis commented on static/src/javascripts/projects/common/modules/ui/lazy-load-images.js:
```diff
@@ -0,0 +1,46 @@
+    const observer = lozad(selector, {
+        rootMargin: '100px 0px', // load images before they are actually viewed
+        load: element => {
+            const isIE =
```
I'm assuming this function block is run for each image? Could we move this `isIE` declaration out of this scope and up into the `lazyLoadImages` scope so we just run this line of code once? We only need to make this declaration once as it'll always be the same.

SiAdcock commented:
In principle I'm on board with this idea, it makes the site much more accessible to people on low bandwidth connections or with data limitations.

I agree we should apply this to fronts first, where the tendency for image bloat is highest, and then consider articles. Some of the most pernicious examples of oversized images have come from onward containers, so lazy-loading these images would be a win.

Since it's so small, we should probably inline the relevant `lozad` code to keep our dependencies down and reduce security attack vectors. 

We should follow up on @JonNorman's revelation during performance hack week presentations that some ads are lazy-loaded in a similar way. Can we consolidate our efforts?

Finally, I can see a few edge cases in which this could potentially lead to a worse experience:

- user disables JavaScript using the `#nojs` URL hash
- standard JavaScript fails to load due for whatever reason
- polyfill.io fails to respond causing standard JavaScript to break for an older browser

However, the vast majority of users will benefit, which I think justifies this change, its just good to be aware of the trade-offs of this approach.

JonNorman commented:
As a follow-up, you're probably best speaking to someone from the commercial team about this as I haven't looked at it in months (@katebee, @janua , @jeteve )but it used to be that the bulk of the code was in [lazy-load.js]( https://github.com/guardian/frontend/blob/master/static/src/javascripts/projects/commercial/modules/dfp/lazy-load.js) (and orchestrated/enabled [here](https://github.com/guardian/frontend/blob/master/static/src/javascripts/projects/commercial/modules/dfp/display-lazy-ads.js)).

On @SiAdcock's final points - these weren't considerations we had to make since our ads won't load if there is no javascript, lazy or otherwise.

Great work by the way!

sndrs commented:
you might want to check the impact of this on scrolling, especially on older devices? 

they used to lazy load, but we found it was killing older android devices so we [removed it](#20809). the availability of `IntersectionObserver` may mean that's no longer a concern, i'm not sure. 

but that tti jump from 18 to 20s+ suggests it might be worth looking into

paperboyo commented:
Aside: not sure of the status of Chrome native lazyload images flag in the wild, but just adding a `lazyload="on"` attribute shouldn’t hurt (explained at the bottom of [this](https://medium.com/@filipvitas/lazy-load-images-with-zero-javascript-2c5bcb691274)). May (or may not) be a cheap way of measuring the impact of this. And won’t brake when js brakes.

SiAdcock commented:
> they used to lazy load, but we found it was killing older android devices so we removed it. the availability of IntersectionObserver may mean that's no longer a concern, i'm not sure.

I've got a pretty ancient Android phone, so we should test the scrolling experience. There's probably a few more in the devices cabinet.

markjamesbutler commented:
@nicl What is the status of this PR? I'm interested to see how this works in the wild and if we are close to merging or not.

nicl commented:
@markjamesbutler we have a fun week next week so I'm going to look at this then. My hope is we can still get it merged, but there is a bit more work to do on it. Partly it is finding the time to do a careful release and some checks to gain more confidence it won't cause significant issues.

markjamesbutler commented:
@markjamesbutler Thank you for the update Nic, good luck with the release.

nicl commented:
@GHaberis have looked at impact on initial doc size:

![screenshot 2019-01-29 at 15 47 38](https://user-images.githubusercontent.com/858402/51921678-c584ba80-23df-11e9-8279-d6850538f53f.png)
![screenshot 2019-01-29 at 16 05 27](https://user-images.githubusercontent.com/858402/51921679-c584ba80-23df-11e9-8172-bc4050f50a23.png)

(On CODE) we see an

nicl commented:
Update:

* have inlined lozad
* added a 0% ab test to reduce risk of initial deploy and ease fuller testing in PROD across devices
* have updated noscript behaviour compared to original PR (to keep the picture element/responsive source behaviour)
* added some fade-in CSS to improve UX

@SiAdcock @GHaberis would be great to get updated thoughts. The aim would be to get this live as soon as possible and do further testing on PROD prior to a larger participation group and eventually full rollout.
